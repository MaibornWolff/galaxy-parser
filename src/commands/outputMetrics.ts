import fs from "fs";
import { Readable } from "stream";
import zlib from "zlib";

interface OutputNode {
    name: string;
    type: "File";
    metrics: { [key: string]: number };
}

interface OutputRelationship {
    from: string;
    to: string;
    metrics: {
        coupling: 100.0;
    };
}

export function outputAsJson(
    fileMetrics: Map<string, Map<string, MetricResult>>,
    relationshipMetrics: CouplingMetricResult,
    outputFilePath: string,
    compress: boolean
) {
    const output = buildOutputObject(fileMetrics, relationshipMetrics);
    const outputString = JSON.stringify(output).toString();

    if (compress) {
        dumpCompressed(
            outputString,
            outputFilePath.endsWith(".gz") ? outputFilePath : outputFilePath + ".gz"
        );
    } else {
        fs.writeFileSync(outputFilePath, outputString);
    }

    console.log("Results saved to " + outputFilePath);
}

function buildOutputObject(
    fileMetrics: Map<string, Map<string, MetricResult>>,
    relationshipMetrics: CouplingMetricResult
) {
    const output: { nodes: OutputNode[]; relationships: OutputRelationship[] } = {
        nodes: [],
        relationships: [],
    };

    for (const [filePath, metricsMap] of fileMetrics.entries()) {
        const metrics: { [key: string]: number } = {};

        for (const [metricName, metricValue] of metricsMap.entries()) {
            metrics[metricName] = metricValue.metricValue;
        }

        output.nodes.push({
            name: filePath,
            type: "File",
            metrics: metrics,
        });
    }

    const couplingMetricResults = relationshipMetrics.metricValue || [];

    for (const couplingMetricResult of couplingMetricResults) {
        output.relationships.push({
            from: couplingMetricResult.fromSource,
            to: couplingMetricResult.toSource,
            metrics: { coupling: 100.0 },
        });
    }

    return output;
}

function dumpCompressed(outputString, outputFilePath) {
    const readableStream = new Readable();
    readableStream.push(outputString);
    readableStream.push(null);

    const gzip = zlib.createGzip();
    const writeStream = fs.createWriteStream(outputFilePath);
    readableStream.pipe(gzip).pipe(writeStream);
}
