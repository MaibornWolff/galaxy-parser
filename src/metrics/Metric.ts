interface MetricResult {
    metricName: string;
    metricValue: number;
}

interface CouplingMetricValue {
    fromNamespace: string, toNamespace: string, fromSource: string, toSource: string
}

interface CouplingMetricResult {
    metricName: string;
    metricValue: CouplingMetricValue[];
}

interface Metric {
    calculate(parseFile: ParseFile): MetricResult;
    getName(): string;
}

interface CouplingMetric {
    calculate(files: ParseFile[]): CouplingMetricResult;
    getName(): string;
}

interface ParseFile {
    language: string;
    filePath: string;
}
