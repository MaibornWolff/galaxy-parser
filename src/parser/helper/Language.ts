import CSharp from "tree-sitter-c-sharp";
import CPlusPlus from "tree-sitter-cpp";
import GO from "tree-sitter-go";
import Java from "tree-sitter-java";
import JavaScript from "tree-sitter-javascript";
import Kotlin from "tree-sitter-kotlin";
import PHP from "tree-sitter-php";
import Python from "tree-sitter-python";
import TypeScript from "tree-sitter-typescript";
import Ruby from "tree-sitter-ruby";
import Rust from "tree-sitter-rust";
import Bash from "tree-sitter-bash";
import C from "tree-sitter-c";
import JSON from "tree-sitter-json";
import { ConstantTwoWayMap } from "./ConstantTwoWayMap";
import { Configuration } from "../Configuration";
import { getFileExtension, lookupLowerCase, replaceForwardWithBackwardSlashes } from "./Helper";
import path from "path";

/**
 * Enum of all supported programming languages.
 */
export const enum Language {
    CSharp,
    CPlusPlus,
    Go,
    Java,
    JavaScript,
    Kotlin,
    PHP,
    TypeScript,
    TSX,
    Python,
    Ruby,
    Rust,
    Bash,
    C,
    JSON,
}

/**
 * Maps back and forth between language and an abbreviation as string value.
 * The string value corresponding to the language is equivalent to the abbreviation used in the node mappings file.
 *
 * Note that this is not necessarily identical to the file extension of a file of the corresponding language,
 * as there can be multiple file extensions for one language. Use {@link fileExtensionToLanguage}
 * or {@link fileExtensionToGrammar} for mapping file extensions.
 */
export const languageToAbbreviation = new ConstantTwoWayMap<Language, string>(
    new Map([
        [Language.CSharp, "cs"],
        [Language.CPlusPlus, "cpp"],
        [Language.Go, "go"],
        [Language.Java, "java"],
        [Language.JavaScript, "js"],
        [Language.Kotlin, "kt"],
        [Language.PHP, "php"],
        [Language.TypeScript, "ts"],
        [Language.TSX, "tsx"],
        [Language.Python, "py"],
        [Language.Ruby, "rb"],
        [Language.Rust, "rs"],
        [Language.Bash, "sh"],
        [Language.C, "c"],
        [Language.JSON, "json"],
    ])
);

/**
 * Maps from a language to the corresponding tree-sitter grammar.
 */
export const languageToGrammar = new Map([
    [Language.CSharp, CSharp],
    [Language.CPlusPlus, CPlusPlus],
    [Language.Go, GO],
    [Language.Java, Java],
    [Language.JavaScript, JavaScript],
    [Language.Kotlin, Kotlin],
    [Language.PHP, PHP.php],
    [Language.TypeScript, TypeScript.typescript],
    [Language.TSX, TypeScript.tsx],
    [Language.Python, Python],
    [Language.Ruby, Ruby],
    [Language.Rust, Rust],
    [Language.Bash, Bash],
    [Language.C, C],
    [Language.JSON, JSON],
]);

/**
 * Maps supported file extensions to the corresponding programming languages.
 * In lower case. This list is for file extensions which are not case-sensitive.
 */
const fileExtensionToLanguage = new Map([
    ["cs", Language.CSharp],
    ["cpp", Language.CPlusPlus],
    ["cp", Language.CPlusPlus],
    ["cxx", Language.CPlusPlus],
    ["cc", Language.CPlusPlus],
    ["hpp", Language.CPlusPlus],
    ["hxx", Language.CPlusPlus],
    ["hh", Language.CPlusPlus],
    ["go", Language.Go],
    ["java", Language.Java],
    ["js", Language.JavaScript],
    ["kt", Language.Kotlin],
    ["php", Language.PHP],
    ["ts", Language.TypeScript],
    ["tsx", Language.TSX],
    ["py", Language.Python],
    ["rb", Language.Ruby],
    ["rs", Language.Rust],
    ["sh", Language.Bash],
    ["json", Language.JSON],
]);

/**
 * Maps supported file extensions to the corresponding programming languages.
 * For case-sensitive file extensions.
 */
const caseSensitiveFileExtensionToLanguage = new Map([
    ["c", Language.C],
    ["C", Language.CPlusPlus],
    ["H", Language.CPlusPlus], // lowercase .h has a special treatment
]);

/**
 * Estimates the language of a file based upon the file extension and file path.
 * @param filePath Path to the file, including the file extension.
 * @param config Configuration to apply.
 * @param pathModule ONLY FOR TESTING PURPOSES: overrides the platform-specific path module.
 */
export function assumeLanguageFromFilePath(
    filePath: string,
    config: Configuration,
    pathModule = path,
) {
    const fileExtension: string = getFileExtension(filePath);

    if (fileExtension === "h") {
        if (shouldHBeParsedAsC(filePath, config, pathModule)) {
            return Language.C;
        }
        return Language.CPlusPlus;
    }

    const resultCaseSensitive = caseSensitiveFileExtensionToLanguage.get(fileExtension);
    if (resultCaseSensitive !== undefined) {
        return resultCaseSensitive;
    }
    return lookupLowerCase(fileExtensionToLanguage, fileExtension);
}

/**
 * Handling of the parse .h as C option.
 */
function shouldHBeParsedAsC(filePath: string, config: Configuration, pathModule = path): boolean {
    if (config.parseAllHAsC) {
        return true;
    }
    if (config.parseSomeHAsC.size > 0) {
        // Use the path relative to the sources path to avoid the unintuitive behaviour
        // that higher-level folders are evaluated for this:
        const relativePath = pathModule.relative(config.sourcesPath, filePath);
        const backwardSlashRelpath = replaceForwardWithBackwardSlashes(relativePath);
        const relpathSplitted = backwardSlashRelpath.split("\\");
        for (const pathElement of relpathSplitted) {
            if (config.parseSomeHAsC.has(pathElement)) {
                return true;
            }
        }
    }
    return false;
}

/**
 * Maps supported file extensions to the corresponding language grammar.
 * @param fileExtension The file extension to map.
 * @return The corresponding grammar if the file extension is supported, undefined otherwise.
 */
export function fileExtensionToGrammar(fileExtension: string) {
    const language = fileExtensionToLanguage.get(fileExtension);
    if (language !== undefined) {
        return languageToGrammar.get(language);
    } else {
        return undefined;
    }
}
