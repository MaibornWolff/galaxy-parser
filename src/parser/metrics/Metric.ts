/**
 * Interface for carrying the result of a metric calculation.
 */
export interface MetricResult {
    /**
     * Name of the metric.
     */
    metricName: string;

    /**
     * Value of the metric.
     */
    metricValue: number;
}

export interface Relationship {
    fromNamespace: string;
    toNamespace: string;
    fromSource: string;
    toSource: string;
    fromClassName: string;
    toClassName: string;
    usageType: string | "usage" | "extends" | "implements";
}

export interface CouplingMetrics {
    outgoing_dependencies: number;
    incoming_dependencies: number;
    coupling_between_objects: number;
    instability: number;
}

export interface CouplingResult {
    relationships: Relationship[];
    metrics: Map<string, CouplingMetrics>;
}

/**
 * Interface for calculating a concrete metric.
 */
export interface Metric {
    /**
     * Calculates the metric value for the supplied file.
     * @param parseFile Source code file for which the metric value should be calculated.
     * @return A MetricResult containing the calculated metric value.
     */
    calculate(parseFile: ParseFile): MetricResult;

    /**
     * Returns the name of this metric.
     * @return The name of this metric.
     */
    getName(): string;
}

export interface CouplingMetric {
    calculate(files: ParseFile[]): CouplingResult;
    getName(): string;
}

/**
 * Represents a file to be parsed, including its path and the file extension.
 */
export interface ParseFile {
    /**
     * File extension of the file.
     */
    fileExtension: string;

    /**
     * Path to the file.
     */
    filePath: string;
}
