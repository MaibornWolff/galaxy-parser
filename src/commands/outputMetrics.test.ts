import { afterAll, beforeEach, describe, expect, it, vi } from "vitest";
import { outputAsJson } from "./outputMetrics.js";
import { MetricResult, CouplingResult, MetricError } from "../parser/metrics/Metric.js";
import { FileType } from "../parser/helper/Language.js";
import { mockConsole } from "../../test/metric-end-results/TestHelper.js";

const writeFileSync = vi.hoisted(() => vi.fn());
vi.mock("fs", async (importOriginal) => {
    const actual = await importOriginal<typeof import("fs")>();
    return { ...actual, writeFileSync };
});
describe("outputMetrics", () => {
    describe("writes json into file ", () => {
        beforeEach(() => {
            mockConsole();
        });

        afterAll(() => {
            vi.resetModules();
        });

        it("when metrics are present", () => {
            const file1MetricResults: MetricResult[] = [];
            file1MetricResults.push({ metricName: "metric1", metricValue: 42 });
            file1MetricResults.push({ metricName: "metric2", metricValue: 43 });

            const file2MetricResults: MetricResult[] = [];
            file2MetricResults.push({ metricName: "metric1", metricValue: 44 });
            const file2MetricErrors: MetricError[] = [];
            file2MetricErrors.push({ metricName: "metric2", error: new Error("Buh!") });

            const fileMetrics = new Map([
                [
                    "/file/path1.test",
                    {
                        fileType: FileType.SourceCode,
                        metricResults: file1MetricResults,
                        metricErrors: [],
                    },
                ],
                [
                    "/file/path2.test",
                    {
                        fileType: FileType.SourceCode,
                        metricResults: file2MetricResults,
                        metricErrors: file2MetricErrors,
                    },
                ],
            ]);

            const unknownFiles = ["/file/path3.unknown", "/file/noExtension"];
            const errorFiles = ["/file/path4.error"];

            const relationShipMetrics: CouplingResult = {
                relationships: [
                    {
                        fromNamespace: "fromNamespace",
                        toNamespace: "toNamespace",
                        usageType: "usage",
                        fromSource: "/file/path2.test",
                        toSource: "/file/path1.test",
                        fromClassName: "ClassA",
                        toClassName: "ClassB",
                    },
                ],
                metrics: new Map([
                    [
                        "/file/path2.test",
                        {
                            outgoing_dependencies: 3,
                            incoming_dependencies: 2,
                            instability: 0.6,
                            coupling_between_objects: 2,
                        },
                    ],
                ]),
            };

            outputAsJson(
                fileMetrics,
                unknownFiles,
                errorFiles,
                relationShipMetrics,
                "mocked-file.json",
                false,
            );

            expect(console.log).toHaveBeenCalledTimes(1);
            expect(console.log).toHaveBeenCalledWith("Results saved to mocked-file.json");

            expect(writeFileSync).toHaveBeenCalledTimes(1);
            const [file, data, options] = writeFileSync.mock.lastCall!;

            expect(file).toBe("mocked-file.json");
            expect(data).toMatchSnapshot();
            expect(options).toBeUndefined();
        });

        it("when no metrics are present", () => {
            const relationShipMetrics: CouplingResult = {
                relationships: [],
                metrics: new Map(),
            };

            outputAsJson(new Map(), [], [], relationShipMetrics, "mocked-file.json", false);

            expect(console.log).toHaveBeenCalledTimes(1);
            expect(console.log).toHaveBeenCalledWith("Results saved to mocked-file.json");

            expect(writeFileSync).toHaveBeenCalledTimes(1);
            expect(writeFileSync).toHaveBeenCalledWith(
                "mocked-file.json",
                '{"nodes":[],"info":[],"relationships":[]}',
            );
        });
    });
});
