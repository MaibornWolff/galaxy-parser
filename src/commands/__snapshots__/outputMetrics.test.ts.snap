// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`outputMetrics writes json into file  when metrics are present 1`] = `"{"nodes":[{"name":"/file/path1.test","type":"source_code","metrics":{"metric1":42,"metric2":43}},{"name":"/file/path2.test","type":"source_code","metrics":{"metric1":44,"outgoing_dependencies":3,"incoming_dependencies":2,"instability":0.6}}],"info":[{"name":"/file/path3.unknown","type":"unsupported_file","message":"Unknown language or file extension"},{"name":"/file/noExtension","type":"unsupported_file","message":"Unknown language or file extension"},{"name":"/file/path4.error","type":"error_file","message":"Error while parsing a syntax tree for the file"},{"name":"/file/path2.test","type":"error_file","message":"Error while calculating the following metric(s) for the file: metric2"}],"relationships":[{"from":"/file/path2.test","to":"/file/path1.test","metrics":{"coupling":100}}]}"`;
